source/as-promise/core.ts

```diff
@@ -0,0 +1,139 @@
+	}
+
+	if (!knownBodyTypes.includes(responseType)) {
+		throw new TypeError(`Unknown body type '${responseType as string}'`);
+	}
+};
+
+export default class PromisableRequest extends Request {
+	['constructor']: typeof PromisableRequest;
+	declare options: NormalizedOptions;
+	declare _throwHttpErrors: boolean;
+
+	static normalizeArguments(url?: string | URL, nonNormalizedOptions?: Options, defaults?: Defaults): NormalizedOptions {
+		const options = super.normalizeArguments(url, nonNormalizedOptions, defaults) as NormalizedOptions;
+
+		assert.any([is.boolean, is.undefined], options.resolveBodyOnly);
+		assert.any([is.boolean, is.undefined], options.methodRewriting);
+		assert.any([is.boolean, is.undefined], options.isStream);
+
+		// `options.retry`
+		const {retry} = options;
+
+		if (defaults) {
+			options.retry = {...defaults.retry};
+		} else {
+			options.retry = {
+				calculateDelay: retryObject => retryObject.computedValue,
+				limit: 0,
+				methods: [],
+				statusCodes: [],
+				errorCodes: [],
+				maxRetryAfter: undefined
+			};
+		}
+
+		if (is.object(retry)) {
+			options.retry = {
+				...options.retry,
+				...retry
+			};
+		} else if (is.number(retry)) {
+			options.retry.limit = retry;
+		}
+
+		if (is.undefined(options.retry.maxRetryAfter)) {
+			options.retry.maxRetryAfter = Math.min(
+				...[options.timeout.request, options.timeout.connect].filter(is.number)
+			);
+		}
+
+		// `options._pagination`
+		if (is.object(options._pagination)) {
+			if (defaults) {
+				(options as Options)._pagination = {
+					...defaults._pagination,
+					...options._pagination
+				};
+			}
+
+			const {_pagination: pagination} = options;
+
+			if (!is.function_(pagination.transform)) {
+				throw new Error('`options._pagination.transform` must be implemented');
+			}
+
+			if (!is.function_(pagination.shouldContinue)) {
+				throw new Error('`options._pagination.shouldContinue` must be implemented');
+			}
+
+			if (!is.function_(pagination.filter)) {
+				throw new TypeError('`options._pagination.filter` must be implemented');
+			}
+
+			if (!is.function_(pagination.paginate)) {
+				throw new Error('`options._pagination.paginate` must be implemented');
+			}
+		}
+
+		return options;
+	}
+
+	static mergeOptions(...sources: Options[]): NormalizedOptions {
+		let mergedOptions: NormalizedOptions | undefined;
+
+		for (const source of sources) {
+			mergedOptions = PromisableRequest.normalizeArguments(undefined, source, mergedOptions);
+		}
+
+		return mergedOptions!;
+	}
+
+	async _beforeError(error: RequestError): Promise<void> {
+		const isHTTPError = error instanceof HTTPError;
+
+		try {
+			for (const hook of this.options.hooks.beforeError) {
+				// eslint-disable-next-line no-await-in-loop
+				error = await hook(error);
+			}
+		} catch (error_) {
+			this.destroy(error_);
+			return;
+		}
+
+		if (this._throwHttpErrors && !isHTTPError) {
+			this.destroy(error);
+		} else {
+			this.emit('error', error);
+		}
```

Comment:
We cannot destroy this if the `retry` functionality is enabled. This is the responsibility of the Got promise.